### PR TITLE
SONAR-6612 ws custom_measures/metrics list available custom metrics for a project

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  public static final String FIELD_DIRECTION = "direction";
+  public static final String FIELD_QUALITATIVE = "qualitative";
+  public static final String FIELD_HIDDEN = "hidden";
+  public static final String FIELD_CUSTOM = "custom";
+  private static final Set<String> OPTIONAL_FIELDS = ImmutableSet.of(FIELD_NAME, FIELD_DESCRIPTION, FIELD_DOMAIN, FIELD_DIRECTION, FIELD_QUALITATIVE, FIELD_HIDDEN, FIELD_CUSTOM);
+  private static final Set<String> MANDATORY_FIELDS = ImmutableSet.of(FIELD_ID, FIELD_KEY, FIELD_TYPE);
+  public static Set<String> ALL_FIELDS = Sets.union(MANDATORY_FIELDS, OPTIONAL_FIELDS);
+
+  {
+    ALL_FIELDS = newHashSet(OPTIONAL_FIELDS);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java]: CRITICAL Synchronize this lazy initialization of 'ALL_FIELDS' [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2444)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  public static final String FIELD_DESCRIPTION = "description";
+  public static final String FIELD_DOMAIN = "domain";
+  public static final String FIELD_TYPE = "type";
+  public static final String FIELD_DIRECTION = "direction";
+  public static final String FIELD_QUALITATIVE = "qualitative";
+  public static final String FIELD_HIDDEN = "hidden";
+  public static final String FIELD_CUSTOM = "custom";
+  private static final Set<String> OPTIONAL_FIELDS = ImmutableSet.of(FIELD_NAME, FIELD_DESCRIPTION, FIELD_DOMAIN, FIELD_DIRECTION, FIELD_QUALITATIVE, FIELD_HIDDEN, FIELD_CUSTOM);
+  private static final Set<String> MANDATORY_FIELDS = ImmutableSet.of(FIELD_ID, FIELD_KEY, FIELD_TYPE);
+  public static Set<String> ALL_FIELDS = Sets.union(MANDATORY_FIELDS, OPTIONAL_FIELDS);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java]: MAJOR Make this "public static ALL_FIELDS" field final [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1444)
MINOR Rename this field "ALL_FIELDS" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS3008)
MAJOR Make ALL_FIELDS a static final constant or non-public and provide accessors if needed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java]: MAJOR Move the contents of this initializer to a standard constructor or to field initializers. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1171)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues:
* CRITICAL 1 critical
* MAJOR 7 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/core/metric/db' package (org.codehaus.sonar:sonar-core:src/main/java/org/sonar/core/metric/db) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/measure/custom/ws' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/measure/custom/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/metric/persistence' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/metric/persistence) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/metric/ws' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/metric/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/MetricsAction.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    this.dbClient = dbClient;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION)
+      .setSince("5.2")
+      .setInternal(true)
+      .setHandler(this)
+      // .setResponseExample(Resources.getResource(getClass(), "example-metrics.json"))
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/MetricsAction.java]: Uncomment this line and provide a valid ``example-metrics.json`` :)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java (hunk 3 above)
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java]: Actually you can get rid of this initialization block altogether, and make ``ALL_FIELDS`` final.

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      writeField(json, FIELD_QUALITATIVE, metric.isQualitative(), fieldsToReturn);
+      writeField(json, FIELD_HIDDEN, metric.isHidden(), fieldsToReturn);
+      writeField(json, FIELD_CUSTOM, metric.isUserManaged(), fieldsToReturn);
+      json.endObject();
+    }
+    json.endArray();
+  }
+
+  private static void writeField(JsonWriter json, String field, String value, Set<String> desiredFields) {
+    if (desiredFields.contains(field)) {
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java]: If you rebase, you'll be able to use ``JsonWriterUtils`` methods for this.

File: server/sonar-server/src/main/resources/org/sonar/server/measure/custom/ws/example-metrics.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (jblievremont) [inline server/sonar-server/src/main/resources/org/sonar/server/measure/custom/ws/example-metrics.json]: See my comment on action class ;)

File: server/sonar-server/src/test/resources/org/sonar/server/measure/custom/ws/MetricsActionTest/metrics.json [hunk 7]
```diff
@@ -0,0 +1,34 @@
+{
+  "metrics": [
+    {
+      "key": "custom-key-1",
+      "name": "custom-name-1",
+      "type": {
+        "key": "INT",
+        "name": "Integer"
```
Comment (jblievremont) [inline server/sonar-server/src/test/resources/org/sonar/server/measure/custom/ws/MetricsActionTest/metrics.json]: Did you change your mind compared to what we discussed yesterday over the response format?

**Comment by jblievremont** [comment]
Please apply my feedback :)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java (hunk 3 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java]: yep, I forgot to delete this code

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/MetricsAction.java (hunk 4 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/MetricsAction.java]: right, thank you

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java (hunk 5 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/MetricJsonWriter.java]: yes

File: server/sonar-server/src/test/resources/org/sonar/server/measure/custom/ws/MetricsActionTest/metrics.json (hunk 7 above)
Comment (teryk) [inline server/sonar-server/src/test/resources/org/sonar/server/measure/custom/ws/MetricsActionTest/metrics.json]: nope, missing test. I should leave my work with a missing test : )

**Comment by jblievremont** [comment]
LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 4 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/core/metric/db' package (org.codehaus.sonar:sonar-core:src/main/java/org/sonar/core/metric/db) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/measure/custom/ws' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/measure/custom/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/metric/persistence' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/metric/persistence) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/metric/ws' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/metric/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

**Pull request merged**